x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/ingest/GeoTileProcessor.java

```diff
@@ -0,0 +1,421 @@
+                if (parent != null && parent.length() > 0) {
+                    ingestDocument.setFieldValue(config.parentField, parent);
+                }
+            }
+            if (config.childrenField != null && config.childrenField.length() > 0) {
+                final String[] children = handler.makeChildren();
+                if (children.length > 0) {
+                    ingestDocument.setFieldValue(config.childrenField, children);
+                }
+            }
+            if (config.nonChildrenField != null && config.nonChildrenField.length() > 0) {
+                final String[] nonChildren = handler.makeNonChildren();
+                if (nonChildren.length > 0) {
+                    ingestDocument.setFieldValue(config.nonChildrenField, nonChildren);
+                }
+            }
+            if (config.depthField != null && config.depthField.length() > 0) {
+                final int depth = handler.getDepth();
+                if (depth >= 0) {
+                    ingestDocument.setFieldValue(config.depthField, depth);
+                }
+            }
+        } catch (Exception e) {
+            throw new IllegalArgumentException("invalid tile definition", e);
+        }
+
+        return ingestDocument;
+    }
+
+    /**
+     * This class handles the different tile types, rectangular or hexagonal.
+     * Each of geohash, geotile and geohex have different ways of describing polygons, and other attributes.
+     */
+    abstract static class TileHandler {
+        Geometry makePolygonFrom(Rectangle r) {
+            double[] x = new double[] { r.getMinX(), r.getMaxX(), r.getMaxX(), r.getMinX(), r.getMinX() };
+            double[] y = new double[] { r.getMinY(), r.getMinY(), r.getMaxY(), r.getMaxY(), r.getMinY() };
+            Polygon polygon = new Polygon(new LinearRing(x, y));
+            if (GeometryNormalizer.needsNormalize(Orientation.CCW, polygon) == false) {
+                return polygon;
+            }
+            return GeometryNormalizer.apply(Orientation.CCW, polygon);
+        }
+
+        abstract Geometry makePolygon();
+
+        abstract String[] makeChildren();
+
+        // Most tilers have no intersecting non-children
+        String[] makeNonChildren() {
+            return new String[] {};
+        }
+
+        abstract int getDepth();
+
+        abstract String getParent();
+    }
+
+    static class GeotileHandler extends TileHandler {
+        private final String address;
+        private final int zoom;
+        private final int x;
+        private final int y;
+
+        GeotileHandler(String address) {
+            this.address = address;
+            final int[] hashAsInts = GeoTileUtils.parseHash(address);
+            this.zoom = hashAsInts[0];
+            this.x = hashAsInts[1];
+            this.y = hashAsInts[2];
+        }
+
+        @Override
+        Geometry makePolygon() {
+            return makePolygonFrom(GeoTileUtils.toBoundingBox(address));
```

Comment:
I think we should index it as a BBOX not as a polygon. The reason is that if we ever support spherical coordinates, the the Polygon would not be equivalent as the tile so we should produce BBOX (Rectangle geometries) instead.

Reply:
Done